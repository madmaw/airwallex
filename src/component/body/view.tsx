import * as React from 'react';
import { Observer } from 'rxjs';
import styles from './view.module.scss';

export type EventBodyRequestInvite = {
  type: 'body_request_invite',
};

export type Events = EventBodyRequestInvite;

export const BUTTON_TEXT_REQUEST = 'Requst an invite';

export function Body({
  observer,
}: {
  observer: Observer<Events>,
}) {
  const requestInvite = React.useCallback(() => {
    observer.next({
      type: 'body_request_invite',
    });
  }, [])

  return (
    <div className={styles.container}>
      <div className={styles.content}>
        <h1>A better way<br/>to enjoy every day.</h1>
        <p>
          Be the first to know when we launch.
        </p>
        <button onClick={requestInvite} autoFocus={true}>{BUTTON_TEXT_REQUEST}</button>
      </div>
    </div>
  );
}