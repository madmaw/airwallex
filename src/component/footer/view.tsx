import * as React from 'react';
import styles from './view.module.scss';

export function Footer() {
  return (
    <div className={styles.container}>
    Made with ❤ in Sydney<br/>
    © 2024 Brocolli & Co. All rights reserved.
    </div>
  )
}