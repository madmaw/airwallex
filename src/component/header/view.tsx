import * as React from 'react';
import styles from './view.module.scss';

export function Header({

}: {

}) {
  return (
    <div className={styles.container}>
      <h1>Broccoli & Co.</h1>
    </div>
  );
}