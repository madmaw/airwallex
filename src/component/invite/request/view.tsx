import * as React from 'react';
import { Observer } from 'rxjs';
import styles from './view.module.scss';

export const PLACEHOLDER_FULL_NAME = 'Full Name';
export const PLACEHOLDER_EMAIL = 'Email';
export const PLACEHOLDER_CONFIRM_EMAIL = 'Confirm Email';
export const BUTTON_TEXT_SEND = 'Send';

export type EventInviteRequestSubmit = {
  type: 'invite_request_submit',
};

export type EventInviteRequestInput = {
  type: 'invite_request_input',
  state: Pick<InviteRequestState, 'email' | 'name' | 'confirmEmail'>,
};

export type Events = 
  | EventInviteRequestSubmit
  | EventInviteRequestInput;

export type InviteRequestState = {
  name?: string,
  email?: string,
  confirmEmail?: string,
  errorMessage?: string,
  processing: boolean,
};

export function InviteRequest({
  state,
  observer,
}: {
  state: InviteRequestState,
  observer: Observer<Events>, 
}) {
  const {
    email,
    name,
    confirmEmail,
    errorMessage,
    processing,
  } = state;
  const setName = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    observer.next({
      type: 'invite_request_input',
      state: {
        ...state,
        name: e.target.value,
      }
    });
  }, [observer]);
  const setEmail = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    observer.next({
      type: 'invite_request_input',
      state: {
        ...state,
        email: e.target.value,
      }
    });
  }, [observer]);
  const setConfirmEmail = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    observer.next({
      type: 'invite_request_input',
      state: {
        ...state,
        confirmEmail: e.target.value,
      }
    });
  }, [observer]);
  const send = React.useCallback(() => {
    observer.next({
      type: 'invite_request_submit',
    });
  }, [observer]);
  return (
    <div className={styles.container}>
      <h2>Request an invite</h2>
      <input
        onChange={setName}
        value={name || ''}
        placeholder={PLACEHOLDER_FULL_NAME}
        disabled={processing}
        autoFocus={true}
        />
      <input
        onChange={setEmail}
        value={email || ''}
        placeholder={PLACEHOLDER_EMAIL}
        disabled={processing}
        />
      <input
        onChange={setConfirmEmail}
        value={confirmEmail || ''}
        placeholder={PLACEHOLDER_CONFIRM_EMAIL}
        disabled={processing}
        />
      <button
        className={styles.submit}
        onClick={send}
        disabled={processing}
        >{processing ? 'Sending, please wait...' : BUTTON_TEXT_SEND}</button>
      {errorMessage && (
        <div className={styles.error}>
          {errorMessage}
        </div>
      )}
    </div>
  );
}
