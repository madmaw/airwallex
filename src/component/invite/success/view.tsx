import * as React from 'react';
import { Observer } from 'rxjs';
import styles from './view.module.scss';

export type EventInviteSuccessDone = {
  type: 'invite_success_done',
};

export type Events = EventInviteSuccessDone;

export const BUTTON_TEXT_OK = 'OK';

export function InviteSuccess({
  observer,
}: {
  observer: Observer<Events>
}) {
  const onClose = React.useCallback(() => {
    observer.next({
      type: 'invite_success_done',     
    });
  }, [observer])
  return (
    <div className={styles.container}>
      <h2>All done!</h2>
      <p>You will be one of the first to experience Broccoli & Co. when we launch</p>
      <button onClick={onClose} autoFocus={true}>{BUTTON_TEXT_OK}</button>
    </div>
  ); 
}