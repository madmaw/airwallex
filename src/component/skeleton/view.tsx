import * as React from 'react';
import styles from './view.module.scss';
import { Observer } from 'rxjs';

export type EventSkeletonCloseDialog = {
  type: 'skeleton_close_dialog',
};

export type Events = EventSkeletonCloseDialog;

type SkeletonProps = {
  header: JSX.Element,
  footer: JSX.Element,
  body: JSX.Element,
  dialog: JSX.Element | undefined,
  observer: Observer<Events>,
};

export const CLOSE_BUTTON_ARIA_LABEL = 'Close';

export function DesktopSkeleton({
  header,
  footer,
  body,
  dialog,
  observer,
}: SkeletonProps) {
  const onClose = React.useCallback(() => {
    observer.next({
      type: 'skeleton_close_dialog',
    });
  }, [observer])

  const preventClose = React.useCallback((e: React.MouseEvent) => {
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();
  }, []);

  return (
    <div className={styles.container}>
      <div className={styles.header}>
        {header}
      </div>
      <div className={styles.body}>
        {body}
      </div>
      <div className={styles.footer}>
        {footer}
      </div>
      {dialog && (
        <div className={styles.overlay} onClick={onClose}>
          <div className={styles.dialog} onClick={preventClose}>
            {dialog}
            <button 
              className={styles.closeButton}
              onClick={onClose}
              aria-label={CLOSE_BUTTON_ARIA_LABEL}
              >x</button>
          </div>
        </div>
      )}
    </div>
  )
}