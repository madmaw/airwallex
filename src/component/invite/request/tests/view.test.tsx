import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { BUTTON_TEXT_SEND, EventInviteRequestInput, Events, InviteRequest, InviteRequestState, PLACEHOLDER_CONFIRM_EMAIL, PLACEHOLDER_EMAIL, PLACEHOLDER_FULL_NAME } from "../view";
import * as React from 'react';
import { Observer } from 'rxjs';
import { RenderResult, fireEvent, render, screen } from '@testing-library/react';

describe('InviteReqeust', () => {
  let observer: Observer<Events>;
  beforeEach(() => {
    observer = {
      complete: vi.fn(),
      next: vi.fn(),
      error: vi.fn(),
    };
  });
  describe('renders', () => {
    describe.each([
      ['is empty', {
        processing: false,
      }],
      ['shows an errorMessage', {
        processing: false,
        errorMessage: 'bad thing',
      }],
      ['shows processing state', {
        processing: true,
      }],
      ['populates name', {
        processing: false,
        name: 'Bob',
      }],
      ['populates email', {
        processing: false,
        email: 'email@place.tld',
      }],
      ['populates confirmation email', {
        processing: false,
        confirmEmail: 'email@place.tld',
      }],
    ] as [string, InviteRequestState][])('%s', (_, state) => {
      let app: RenderResult;
      beforeEach(() => {
        app = render(<InviteRequest observer={observer} state={state}/>);
      });
      afterEach(() => {
        app.unmount();
      });
  
      it('snapshot', () => {
        expect(
          app.baseElement
        ).toMatchSnapshot();
      });  
    });  
  });
  
  describe('interactivity', () => {
    let app: RenderResult;
    const name = 'Bob';
    const email = 'bob@test.com';
    const confirmEmail = 'bad@test.con';
    const BASE_STATE: Pick<InviteRequestState, 'name' | 'email' | 'confirmEmail'> = {
      name,
      email,
      confirmEmail,  
    };
    beforeEach(() => {
      const state: InviteRequestState = {
        processing: false,
        confirmEmail,
        email,
        name,
      };
      app = render(<InviteRequest observer={observer} state={state}/>);
    });

    afterEach(() => {
      app.unmount();
    });

    it('sends the data when you click', () => {
      const sendButton = app.getByText(BUTTON_TEXT_SEND);
      expect(observer.next).not.toHaveBeenCalled();
      sendButton.click();
      expect(observer.next).toHaveBeenCalledTimes(1);
      expect(observer.next).toHaveBeenCalledWith({
        type: 'invite_request_submit',
      });
    });

    it.each([
      [PLACEHOLDER_FULL_NAME, 'Sam', {
        name: 'Sam',
      }],
      [PLACEHOLDER_EMAIL, 'x@y.z', {
        email: 'x@y.z',
      }],
      [PLACEHOLDER_CONFIRM_EMAIL, 'x@y.z', {
        confirmEmail: 'x@y.z',
      }],
    ] as [string, string, Partial<InviteRequestState>][])(
      'updates the %s field',
      (placeholderText, value, state) => {
        const fullName = app.getByPlaceholderText(placeholderText) as HTMLInputElement;
        fireEvent.change(fullName, {
          target: {
            value,
          },
        });
        expect(observer.next).toHaveBeenCalledTimes(1);
        expect(observer.next).toHaveBeenCalledWith({
          state: expect.objectContaining({
            ...BASE_STATE,
            ...state,
          }),
          type: 'invite_request_input',
        });
      });
  });
});