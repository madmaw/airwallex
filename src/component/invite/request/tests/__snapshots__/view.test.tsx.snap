// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InviteReqeust > renders > is empty > snapshot 1`] = `
<body>
  <div>
    <div
      class="_container_075bbc"
    >
      <h2>
        Request an invite
      </h2>
      <input
        placeholder="Full Name"
        value=""
      />
      <input
        placeholder="Email"
        value=""
      />
      <input
        placeholder="Confirm Email"
        value=""
      />
      <button
        class="_submit_075bbc"
      >
        Send
      </button>
    </div>
  </div>
</body>
`;

exports[`InviteReqeust > renders > populates confirmation email > snapshot 1`] = `
<body>
  <div />
  <div />
  <div />
  <div />
  <div />
  <div>
    <div
      class="_container_075bbc"
    >
      <h2>
        Request an invite
      </h2>
      <input
        placeholder="Full Name"
        value=""
      />
      <input
        placeholder="Email"
        value=""
      />
      <input
        placeholder="Confirm Email"
        value="email@place.tld"
      />
      <button
        class="_submit_075bbc"
      >
        Send
      </button>
    </div>
  </div>
</body>
`;

exports[`InviteReqeust > renders > populates email > snapshot 1`] = `
<body>
  <div />
  <div />
  <div />
  <div />
  <div>
    <div
      class="_container_075bbc"
    >
      <h2>
        Request an invite
      </h2>
      <input
        placeholder="Full Name"
        value=""
      />
      <input
        placeholder="Email"
        value="email@place.tld"
      />
      <input
        placeholder="Confirm Email"
        value=""
      />
      <button
        class="_submit_075bbc"
      >
        Send
      </button>
    </div>
  </div>
</body>
`;

exports[`InviteReqeust > renders > populates name > snapshot 1`] = `
<body>
  <div />
  <div />
  <div />
  <div>
    <div
      class="_container_075bbc"
    >
      <h2>
        Request an invite
      </h2>
      <input
        placeholder="Full Name"
        value="Bob"
      />
      <input
        placeholder="Email"
        value=""
      />
      <input
        placeholder="Confirm Email"
        value=""
      />
      <button
        class="_submit_075bbc"
      >
        Send
      </button>
    </div>
  </div>
</body>
`;

exports[`InviteReqeust > renders > shows an errorMessage > snapshot 1`] = `
<body>
  <div />
  <div>
    <div
      class="_container_075bbc"
    >
      <h2>
        Request an invite
      </h2>
      <input
        placeholder="Full Name"
        value=""
      />
      <input
        placeholder="Email"
        value=""
      />
      <input
        placeholder="Confirm Email"
        value=""
      />
      <button
        class="_submit_075bbc"
      >
        Send
      </button>
      <div
        class="_error_075bbc"
      >
        bad thing
      </div>
    </div>
  </div>
</body>
`;

exports[`InviteReqeust > renders > shows processing state > snapshot 1`] = `
<body>
  <div />
  <div />
  <div>
    <div
      class="_container_075bbc"
    >
      <h2>
        Request an invite
      </h2>
      <input
        disabled=""
        placeholder="Full Name"
        value=""
      />
      <input
        disabled=""
        placeholder="Email"
        value=""
      />
      <input
        disabled=""
        placeholder="Confirm Email"
        value=""
      />
      <button
        class="_submit_075bbc"
        disabled=""
      >
        Sending, please wait...
      </button>
    </div>
  </div>
</body>
`;
